README.md
# Wisconsin Benchmark

[Wisconsin Benchmark](http://jimgray.azurewebsites.net/benchmarkhandbook/chapter4.pdf) is a benchmark for relational database systems and machines developed at University of Wisconsin. It was used to asses early relational data system performance, but it will be useful for the capability of DataFrame and its scalability because it is well designed synthetic dataset.

### Table: Attribute Specification of "Scalable" Wisconsin Benchmark

|Attribute Name| Range of Values | Order | Comment |
|--------------|-----------------|-------|---------|
|unique1       |0-(MAXTUPLES-1)  |random | unique, random order|
|unique2       |0-(MAXTUPLES-1)  |sequential | unique, sequential|
|two           |0-1 |random |(unique1 mod 2)|
|four          |0-3 |random |(unique1 mod 4)|
|ten           |0-9 |random |(unique1 mod 10)|
|twenty        |0-19 |random |(unique1 mod 20)|
|onePercent    |0-99 |random |(unique1 mod 100)|
|tenPercent    |0-9 |random |(unique1 mod 10)|
|twentyPercent |0-4 |random |(unique1 mod 5)|
|fiftyPercent  |0-1 |random |(unique1 mod 2)|
|unique3       |0-(MAXTUPLES-1)| random|unique1|
|evenOnePercent |0,2,4,...,198| random |(onePercent * 2)|
|oddOnePercent |1,3,5,...,199| random |(onePercent * 2)+1|
|stringu1      | - |random |_unique, random order, 52bytes each_|
|stringu2      | - |_sequential_|_unique, sequential, 52bytes each_|
|string4       | - |cyclic|4 _unique string, 52bytes each_|

(This table is taken from Table 2 in [Wisconsin Benchmark](http://jimgray.azurewebsites.net/benchmarkhandbook/chapter4.pdf) and _italic_ letters are added to describe in detail.)

### Benchmark dataset

This project will offer the 'Scaled Wisconsin Benchmark Dataset' in arrow, parquet and csv for 1,000 to 100,000,000 rows.

### Benchmark suites for DataFrames

Coming soon.

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add wisconsin-benchmark

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install wisconsin-benchmark

## Usage

To experiment with that code, run `bin/console` for an interactive prompt.

```ruby
require 'wisconsin-benchmark' # if it is not in `bin/console`

table = WisconsinBenchmark::Table.new(1000).generate

```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test-unit` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/wisconsin-benchmark. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[USERNAME]/wisconsin-benchmark/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Wisconsin::Benchmark project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/wisconsin-benchmark/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2022-12-20

- Initial release


test/wb_test.rb
require 'wisconsin-benchmark'
require 'test_helper'

class WisconsinBenchmarkTest < Test::Unit::TestCase
  test 'VERSION' do
    assert do
      ::WisconsinBenchmark.const_defined?(:VERSION)
    end
  end

  test 'out of range in array' do
    assert_raise(RuntimeError) { WisconsinBenchmark::Array.new(100_000_001) }
  end

  sub_test_case 'numeric array' do
    data(
      '1E1' => [10, [9, 1, 2, 3, 5, 7, 8, 4, 6, 0]],
      '1E2' => [100, [69, 9, 86, 56, 65, 51, 31, 35, 99, 58]],
      '1E3' => [1000, [147, 931, 714, 711, 883, 439, 670, 543, 425, 800]],
      '1E4' => [10000, [8800, 1891, 3420, 9850, 7164, 8009, 5057, 6701, 4321, 3043]],
      '1E5' => [100000, [32293, 9402, 71151, 51373, 13756, 22185, 55231, 53191, 8699, 30916]],
      '1E6' => [1000000,
                [439436, 890983, 297656, 161417, 106705, 828869, 423851, 53484, 692558, 217435]],
      '1E7' => [10000000,
                [44520, 9393930, 2115678, 6407432, 1965797, 4782598, 9126488, 5685530, 9644779,
                 5044722]],
      # '1E8' => [100000000,
      #           [440, 9260, 194480, 4084100, 85766120, 1088414, 22856714, 79990986, 79810614,
      #            76022802]],
    )

    def test_numeric_array(data)
      size, values10 = data
      array = WisconsinBenchmark::Array.new(size)

      assert_instance_of Numo::UInt32, array.unique1
      assert_equal values10, array.unique1[0..9].to_a
      assert_equal 0, array.unique1.min
      assert_equal size - 1, array.unique1.max

      assert_instance_of Numo::UInt32, array.unique2
      assert_equal array.unique2.to_a, array.unique1.sort.to_a
    end
  end

  sub_test_case 'string array' do
    data(
      '1E1' => [10, %w[
        AAAAAAJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAABxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAACxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAADxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAAFxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAAHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAAIxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAAExxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAAGxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      ]],
      '1E2' => [100, %w[
        AAAAACRxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAAJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAADIxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAACExxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAACNxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAABZxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAABFxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAABJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAADVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAACGxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      ]],
      '1E3' => [1_000, %w[
        AAAAAFRxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAABJVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAABBMxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAABBJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAABHZxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAQXxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAZUxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAUXxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAAQJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAABEUxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      ]],
      '1E4' => [10_000, %w[
        AAAANAMxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAACUTxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAFBOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAOOWxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAKPOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAALWBxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAHMNxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAJXTxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAGKFxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAENBxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      ]],
      '1E5' => [100_000, %w[
        AAABVUBxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAANXQxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAEBGPxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAACXZXxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAUJCxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAABGVHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAADDSHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAADARVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAAMWPxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAABTTCxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      ]],
      '1E6' => [1_000_000, %w[
        AAAZABKxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AABYSAPxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAQYIIxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAJEUJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAGBWBxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AABVEDPxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAYCZZxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAADBDCxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AABNKMWxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAAMJQXxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      ]],
      '1E7' => [10_000_000, %w[
        AAACNWIxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAUOMJAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAEQJSGxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAOAOLSxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAEHVZPxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAKMCWCxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AATZGSUxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAMLMOGxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AAVCTLBxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        AALBAPUxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      ]],
      # '1E8' => [100_000_000, %w[
      #   AAAAAQYxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   AAAANSExxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   AAALBSAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   AAIYJOUxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   AHFRSYYxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   AACJYCCxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   ABYALSKxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   AGTBDWKxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   AGSQXBAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #   AGKJJTYxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      # ]],
    )

    def test_string_array(data)
      size, values10 = data
      array = WisconsinBenchmark::Array.new(size)
      min = 'AAAAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'

      assert_instance_of Array, array.stringu1
      assert_equal values10, array.stringu1[0..9].to_a
      assert_equal min, array.stringu1.min

      assert_instance_of Array, array.stringu2
      assert_equal array.stringu2.to_a, array.stringu1.sort.to_a

      expect = %w[
        AAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        HHHHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        OOOOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        VVVVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      ]
      expect = (expect * 3)[0..9]
      assert_instance_of Array, array.string4
      assert_equal expect, array.string4[0..9].to_a
    end
  end

  sub_test_case 'table' do
    test 'generate table' do
      table = WisconsinBenchmark::Table.new(20).generate

      expect = <<~STR
        \tunique1\tunique2\ttwo\tfour\tten\ttwenty\tonePercent\ttenPercent\ttwentyPercent\tfiftyPercent\tunique3\tevenOnePercent\toddOnePercent\tstringu1\tstringu2\tstring4
         0\t      9\t      0\t  1\t   1\t  9\t     9\t         9\t         9\t            4\t           1\t      9\t            18\t           19\tAAAAAAJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         1\t     14\t      1\t  0\t   2\t  4\t    14\t        14\t         4\t            4\t           0\t     14\t            28\t           29\tAAAAAAOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAABxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tHHHHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         2\t     13\t      2\t  1\t   1\t  3\t    13\t        13\t         3\t            3\t           1\t     13\t            26\t           27\tAAAAAANxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAACxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tOOOOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         3\t      1\t      3\t  1\t   1\t  1\t     1\t         1\t         1\t            1\t           1\t      1\t             2\t            3\tAAAAAABxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAADxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tVVVVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         4\t     19\t      4\t  1\t   3\t  9\t    19\t        19\t         9\t            4\t           1\t     19\t            38\t           39\tAAAAAATxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAExxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         5\t      2\t      5\t  0\t   2\t  2\t     2\t         2\t         2\t            2\t           0\t      2\t             4\t            5\tAAAAAACxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAFxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tHHHHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         6\t      3\t      6\t  1\t   3\t  3\t     3\t         3\t         3\t            3\t           1\t      3\t             6\t            7\tAAAAAADxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAGxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tOOOOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         7\t     10\t      7\t  0\t   2\t  0\t    10\t        10\t         0\t            0\t           0\t     10\t            20\t           21\tAAAAAAKxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tVVVVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         8\t      5\t      8\t  1\t   1\t  5\t     5\t         5\t         5\t            0\t           1\t      5\t            10\t           11\tAAAAAAFxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAIxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
         9\t     12\t      9\t  0\t   0\t  2\t    12\t        12\t         2\t            2\t           0\t     12\t            24\t           25\tAAAAAAMxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tHHHHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        10\t      7\t     10\t  1\t   3\t  7\t     7\t         7\t         7\t            2\t           1\t      7\t            14\t           15\tAAAAAAHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAKxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tOOOOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        11\t      8\t     11\t  0\t   0\t  8\t     8\t         8\t         8\t            3\t           0\t      8\t            16\t           17\tAAAAAAIxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAALxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tVVVVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        12\t     18\t     12\t  0\t   2\t  8\t    18\t        18\t         8\t            3\t           0\t     18\t            36\t           37\tAAAAAASxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAMxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        13\t     11\t     13\t  1\t   3\t  1\t    11\t        11\t         1\t            1\t           1\t     11\t            22\t           23\tAAAAAALxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAANxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tHHHHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        14\t      4\t     14\t  0\t   0\t  4\t     4\t         4\t         4\t            4\t           0\t      4\t             8\t            9\tAAAAAAExxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tOOOOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        15\t     15\t     15\t  1\t   3\t  5\t    15\t        15\t         5\t            0\t           1\t     15\t            30\t           31\tAAAAAAPxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAPxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tVVVVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        16\t     17\t     16\t  1\t   1\t  7\t    17\t        17\t         7\t            2\t           1\t     17\t            34\t           35\tAAAAAARxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAAQxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        17\t     16\t     17\t  0\t   0\t  6\t    16\t        16\t         6\t            1\t           0\t     16\t            32\t           33\tAAAAAAQxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAARxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tHHHHxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        18\t      6\t     18\t  0\t   2\t  6\t     6\t         6\t         6\t            1\t           0\t      6\t            12\t           13\tAAAAAAGxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAASxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tOOOOxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        19\t      0\t     19\t  0\t   0\t  0\t     0\t         0\t         0\t            0\t           0\t      0\t             0\t            1\tAAAAAAAxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tAAAAAATxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\tVVVVxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      STR
      assert_equal expect, table.to_s
    end
  end
end


test/benchmark.rb
# frozen_string_literal: true

require 'test_helper'

class WisconsinBenchmarkTest < Test::Unit::TestCase
  include WisconsinBenchmark

  test 'VERSION' do
    assert do
      ::WisconsinBenchmark.const_defined?(:VERSION)
    end
  end

  sub_test_case 'array' do
    test 'something useful' do
      assert_kind_of? Numo::UInt32, WisconsinBenchmark::Array.new
      assert_raise(RuntimeError) { WisconsinBenchmark::Array.new(1E9.to_i) }
    end
  end
end


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'wisconsin-benchmark/table'

require 'test-unit'


lib/wisconsin-benchmark/table.rb
module WisconsinBenchmark
  # Arrow::Table generator
  class Table
    # Create Scaled Wisconsin Benchmark dataset object.
    #
    # @return size [Integer] number of tuples.
    #
    def initialize(size)
      @size = size
      @array = WisconsinBenchmark::Array.new(size)
    end

    attr_reader :table, :size

    # Generate Scaled Wisconsin Benchmark dataset in Arrow::Table.
    #
    # @return [Arrow::Table] generated dataset in Arrow::Table.
    #
    def generate
      @table = Arrow::Table.new(
        [
          [:unique1,        unique1 = @array.unique1],
          [:unique2,        @array.unique2],
          [:two,            unique1 % 2],
          [:four,           unique1 % 4],
          [:ten,            unique1 % 10],
          [:twenty,         unique1 % 20],
          [:onePercent,     onePercent = unique1 % 100],
          [:tenPercent,     unique1 % 10],
          [:twentyPercent,  unique1 % 5],
          [:fiftyPercent,   unique1 % 2],
          [:unique3,        unique1],
          [:evenOnePercent, onePercent * 2],
          [:oddOnePercent,  (onePercent * 2) + 1],
          [:stringu1,       @array.stringu1],
          [:stringu2,       @array.stringu2],
          [:string4,        @array.string4]
        ]
      )
    end
  end
end


lib/wisconsin-benchmark/array.rb
module WisconsinBenchmark
  # Array generator
  class Array
    def initialize(size)
      if    size <= 1000      then @generator = 279;   @prime = 1009
      elsif size <= 10000     then @generator = 2969;  @prime = 10007
      elsif size <= 100000    then @generator = 21395; @prime = 100003
      elsif size <= 1000000   then @generator = 2107;  @prime = 1000003
      elsif size <= 10000000  then @generator = 211;   @prime = 10000019
      elsif size <= 100000000 then @generator = 21;    @prime = 100000007
      else
        raise "too many rows requested #{size}"
      end
      @size = size
    end

    # Create a random/unique record array
    #   0...size in range
    #
    # @return [Numo::UInt32] array of attribute :unique1.
    #
    def unique1
      @unique1 ||= begin
        warn 'Generating unique1'

        seed = @generator
        ary = @size.times.map do
          seed = rand(seed)
          seed - 1
        end
        Numo::UInt32.new(@size).store(ary)
      end
    end

    # Create a sequential record array as 0...size.
    #
    # @return [Numo::UInt32] array of attribute :unique2.
    #
    def unique2
      @unique2 ||= begin
        warn 'Generating unique2'

        Numo::UInt32.new(@size).seq
      end
    end

    # Create a randomly distributed string array, stringu1.
    #
    # @return [Array<String>]
    #   array of generated randomly distributed distinct strings.
    #   - Each string is 52 bytes.
    #   - string[0..6] is converted string from numbers in randomly distributed unique1.
    #     The numbers are converted to n-adic string mapped for 'A-Z'.
    #     {1 => 'AAAAAAA', 2 => 'AAAAAAB', ... 26 => 'AAAAAAZ', 27 => 'AAAAABA', ...}
    #   - string[7..] is 'x' * 45
    # @example
    #  ['AAAAAFRxxxxx ... ', 'AAAABJVxxxxx ... ', 'AAAABBMxxxxx ... ', ... ]
    #
    def stringu1
      @stringu1 ||= begin
        warn 'Generating stringu1'

        trailer = 'x' * 45
        unique1.to_a.map do |i|
          str = i.to_s(26).tr('0-9a-p', 'A-Z')
          ('A' * (7 - str.size)) << str << trailer
        end
      end
    end

    # Create a sequencial string array, stringu2.
    #
    # @return [Array<String>]
    #   sequential array of strings.
    #   - Each string is 52 bytes.
    #   - string[0..6] is sequential string consists of 'A'..'Z' started from 'AAAAAAA'.
    #   - string[7..] is 'x' * 45
    # @example
    #  ['AAAAAAAxxxxx ... ', 'AAAAAABxxxxx ... ', 'AAAAAACxxxxx ... ', ... ]
    #
    def stringu2
      @stringu2 ||= begin
        warn 'Generating stringu2'

        trailer = 'x' * 45
        (('A' * 7)..).take(@size).map { _1 << trailer }
      end
    end

    # Create a cyclic repeated string array, string4.
    #
    # @return [Array<String>]
    #   cyclic repeating array of strings.
    #   - Each string is 52 bytes.
    #   - string[0..3] is one of ['AAAA', 'HHHH', 'OOOO', 'VVVV'].
    #   - string[7..] is 'x' * 48
    #   - four strings are repeated.
    # @example
    #  ['AAAAxxxxx ... ', 'HHHHxxxxx ... ', 'OOOOxxxxx ... ', 'VVVVxxxxx ...', ... ]
    #
    def string4
      @string4 ||= begin
        warn 'Generating string4'

        trailer = 'x' * 48
        array = %w[AAAA HHHH OOOO VVVV].map { _1 << trailer }
        @size.times.map { |i| array[i % 4] }
      end
    end

    private

    # Get pseudo-random integers by a linear congruential generator.
    def rand(seed)
      loop do
        seed = (@generator * seed) % @prime
        break if seed <= @size
      end
      seed
    end
  end
end
